Lab 6 - CRUD puppy with interface

undrewb opened the pull request:
Fixes #532 

Review of colleague's PR #424 

#### Changes proposed in this PR:

Implement a Puppy struct containing ID, Breed, Colour, Value.
Create Storer interface with CRUD methods for Puppy
Write a MapStore implementation of Storer backed by a map
Write a SyncStore implementation of Storer backed by a sync.Map
Keep all implementation files in the same folder and in package main
Test against the Storer interface and run in suite with both implementations

Commit:
Implement CreatePuppy in storer

Create puupy struct
Implement Init to initialize a puppy and assign a random id
Implement String method to print a puppy struct
At this point there is no check for collision ids
Implement a map backed storer only at this time

Implement ReadStore

Error checking hasnt been implemented.
At this stage nil will be returned if the item cant be found.
This will be addressed in a later lab

Implement DeletePuppy
Fix boolean logic in ReadPuppy

Implement UpdatePuppy
Implement clone method on puppy
Resolve problem with usage of pointers and references

Separate MapStore into its own file
Create a basic main_test

Integrate linter feedback

Rename ID to id where used as parameter

Change mapstore test to use a test suite using the Storer interface
Begin implementing a syncstore version of the Storer interface

Implement all methids for syncstore
Implement all cases for store_tests
Ensure 100% test coverqge

Fix linter errors

Capitals and newlines in error messages
Not assign table tests when ranging across them
Misalignment of structs

Verified lint and code cov passed

Separate types into their own file

Resolve linting issue in types.go

willshen8 commented on .gitignore:
```diff
@@ -11,3 +11,4 @@
 # Output of the go coverage tool, specifically when used with LiteIDE
 *.out
 coverage.txt
```
Oops, this file should have not been committed.

willshen8 commented on 06_puppy/undrewb/types.go:
```diff
@@ -0,0 +1,24 @@
+	puppy.Value = p.Value
+	return puppy
+}
+
```
I like the idea of immutability.

willshen8 commented on 06_puppy/undrewb/MapStore.go:
```diff
@@ -0,0 +1,48 @@
+	if puppy == nil {
+		return fmt.Errorf("cant update to a nil puppy")
+	}
+	p, exist := (*msp)[id]
```
This can be rewritten as follow (seem to be the way go is written mostly)
```
if p, exist := (*msp)[id]; !exist {
    return errors.New("no such puppy")
}
```

undrewb replied:
That's different logic. I'm making sure the puppy passed in isn't nil.

willshen8 commented on 06_puppy/undrewb/store_test.go:
```diff
@@ -0,0 +1,177 @@
+			puppy:   corgi,
+			wantErr: true},
+	}
+	for _, tt := range tests {
```
This is just a comment. I have written my test cases as unit test cases, as each one test a specific condition. Rather than bundle all the create operations into one. Not saying it's right or wrong. But I think test cases should be atomic. And it will eliminate the need for struct, and making test files shorter.

willshen8 commented on 06_puppy/undrewb/store_test.go:
```diff
@@ -0,0 +1,177 @@
+				t.Errorf("ReadPuppy() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+			if !reflect.DeepEqual(got, tt.want) {
```
You can try to use the 'assert.Equal' library, and reduce your code length:

e.g. ` assert.Equal(a, b, "The two words should be the same.")`

undrewb replied:
Equal and DeepEqual are quite different.

willshen8 commented on 06_puppy/undrewb/store_test.go:
```diff
@@ -0,0 +1,177 @@
+	s.store = s.storerType()
+
+	corgi := &Puppy{1000, "corgi", "orange and white", "$$$$$$$"}
+	_ = s.store.CreatePuppy(corgi)
```
I don't believe you should have `createPuppy` method inside `SetUpTest` func, as if your `createPuppy` method is wrong, then it would have failed here. Consider just have raw data in here.

undrewb replied:
If I tie the insert to implementation I cant use the same code for testing different implementations of the same interface. If the test fails because the implementation is wrong thats a good thing :)

willshen8 commented on 06_puppy/undrewb/store_test.go:
```diff
@@ -0,0 +1,177 @@
+		},
+		{name: "lookup non-existing puppy",
+			id:      1001,
+			want:    nil,
```
I think this is bit verbose, you could have just run `suite.Nil(expected, actual)` in your test cases.

undrewb replied:
I only want nil for a specific case, not for all tests. When I want corgi (line 79) suite.Nil isn't appropriate.

willshen8 commented on 06_puppy/undrewb/store_test.go:
```diff
@@ -0,0 +1,177 @@
+		s.T().Run(tt.name, func(t *testing.T) {
+			got, err := s.store.DeletePuppy(tt.id)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("DeletePuppy() error = %v, wantErr %v", err, tt.wantErr)
```
Same as above. just use `assert.Equal` to test for differences.

undrewb replied:
Same as above.

willshen8 reviewed (changes requested):
It's mostly pretty good.

1. Uncommit gitignore file.
2. Fix up your test cases.

Commit:
Integrate code review feedback

Change format of test case data
Change parameter ordering of test cases
Implement ID assignment in storers

Commit:
Stop tracking gitignore file

juliaogris commented:
The go-course is now closed. Thank you very much for participating.
